x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java

```diff
@@ -0,0 +1,297 @@
+            {
+                builder.startObject("_meta");
+                builder.field("version", Version.CURRENT.toString());
+                builder.endObject();
+
+                builder.field("dynamic", "strict");
+                builder.startObject("properties");
+                {
+                    builder.startObject(NAME_FIELD.getPreferredName());
+                    builder.field("type", "keyword");
+                    builder.endObject();
+
+                    builder.startObject(INDICES_FIELD.getPreferredName());
+                    builder.field("type", "keyword");
+                    builder.endObject();
+
+                    builder.startObject(BINARY_CONTENT_FIELD.getPreferredName());
+                    builder.field("type", "object");
+                    builder.field("enabled", "false");
+                    builder.endObject();
+                }
+                builder.endObject();
+            }
+            builder.endObject();
+            return builder;
+        } catch (IOException e) {
+            logger.fatal("Failed to build " + ENGINE_CONCRETE_INDEX_NAME + " index mappings", e);
+            throw new UncheckedIOException("Failed to build " + ENGINE_CONCRETE_INDEX_NAME + " index mappings", e);
+        }
+    }
+
+    /**
+     * Gets the {@link Engine} from the index if present, or delegate a {@link ResourceNotFoundException} failure to the provided
+     * listener if not.
+     *
+     * @param resourceName The resource name.
+     * @param listener The action listener to invoke on response/failure.
+     */
+    public void getEngine(String resourceName, ActionListener<Engine> listener) {
+        final GetRequest getRequest = new GetRequest(ENGINE_ALIAS_NAME).id(resourceName).realtime(true);
+        clientWithOrigin.get(getRequest, listener.delegateFailure((delegate, getResponse) -> {
+            if (getResponse.isExists() == false) {
+                delegate.onFailure(new ResourceNotFoundException(resourceName));
+                return;
+            }
+            final BytesReference source = getResponse.getSourceInternal();
+            final Engine res = parseEngineBinaryFromSource(source);
+            delegate.onResponse(res);
+        }));
+    }
+
+    /**
+     * Creates or updates the {@link Engine} in the underlying index.
+     *
+     * @param engine The engine object.
+     * @param listener The action listener to invoke on response/failure.
+     */
+    public void putEngine(Engine engine, ActionListener<IndexResponse> listener) {
+        try (ReleasableBytesStreamOutput buffer = new ReleasableBytesStreamOutput(0, bigArrays.withCircuitBreaking())) {
+            try (XContentBuilder source = XContentFactory.jsonBuilder(buffer)) {
+                source.startObject()
+                    .field(NAME_FIELD.getPreferredName(), engine.name())
+                    .field(INDICES_FIELD.getPreferredName(), engine.indices())
+                    .directFieldAsBase64(
```

Comment:
What I understood from the code is that we'll be storing a field with the base64 object representation so we can check versioning, and perform the serialization / deserialization from this field.

It strikes me as a bit odd to store two representations of the object. Isn't there another simpler way of achieving this goal? How are other system indices doing this?

Reply:
We need the two representations the for transport and rest client anyway. 
I am used to handle BWC in the transport format but it's possible to do it at the rest layer too. I don't think we decided on the rest versioning though. I also anticipated the cost of deserialising in xcontent, it's probably premature but I am concerned since in the first version we'll get the document from the index on every search request.
@kderusso proposed that we discuss this part in a follow up to open the work on the APIs.  We can go back to the rest serialisation later in the branch. 
WDYT?

Reply:
Sounds good to me 👍